# HG changeset patch
# User Mike Hommey <mh+mozilla@glandium.org>
# Date 1721709229 0
# Node ID 8f0a27b175cc061981784876003ea1118ec03ece
# Parent  0d7d37ad97feb84af4711cf11b7e60218df9eb1b
Bug 1907258 - Check that flags are not "empty" before splitting them. r=firefox-build-system-reviewers,ahochheiden

The various *FLAGS variables are defined as nargs=1, which guarantees
that if they are set they are going to have a value. They also come with
a default value of "", which works just fine in the default case.

Things change, though, when the variable is explicitly set to an empty
value, which then makes it a NegativeOptionValue, which is different
from the default and needs to be handled.

Differential Revision: https://phabricator.services.mozilla.com/D217328

diff --git a/build/moz.configure/finalize-flags.configure b/build/moz.configure/finalize-flags.configure
--- a/build/moz.configure/finalize-flags.configure
+++ b/build/moz.configure/finalize-flags.configure
@@ -22,17 +22,18 @@ def os_ldflags(
     android_flags,
     thumb_option,
 ):
     flags = []
     if android_flags:
         flags.extend(android_flags.ldflags)
     if thumb_option:
         flags.extend(thumb_option)
-    flags.extend(split(env_ldflags[0]))
+    if env_ldflags:
+        flags.extend(split(env_ldflags[0]))
     flags.extend(linker_flags.ldflags)
     if linker_ldflags:
         flags.extend(linker_ldflags)
     if pack_relative_relocs_flags:
         flags.extend(pack_relative_relocs_flags)
     return flags
 
 
diff --git a/build/moz.configure/flags.configure b/build/moz.configure/flags.configure
--- a/build/moz.configure/flags.configure
+++ b/build/moz.configure/flags.configure
@@ -564,80 +564,89 @@ def host_cppflags(base_cppflags, host, c
             flags += ["-mwindows"]
         flags += ["-DXP_WIN", "-DWIN32", "-D_WIN32", "-D_CRT_SECURE_NO_WARNINGS"]
         if host.cpu == "x86_64":
             flags += ["-D_AMD64_"]
     elif host.kernel == "Darwin":
         flags += ["-DXP_UNIX", "-DXP_MACOSX"]
     else:
         flags += ["-DXP_UNIX"]
-    flags += split(base_cppflags[0])
+    if base_cppflags:
+        flags += split(base_cppflags[0])
     return flags
 
 
 @depends("HOST_CFLAGS", compilation_flags)
 @imports(_from="mozbuild.shellutil", _import="split")
 def host_cflags(base_cflags, compilation_flags):
     flags = list(compilation_flags.host_cflags)
-    flags += split(base_cflags[0])
+    if base_cflags:
+        flags += split(base_cflags[0])
     return flags
 
 
 @depends("HOST_CXXFLAGS", compilation_flags)
 @imports(_from="mozbuild.shellutil", _import="split")
 def host_cxxflags(base_cxxflags, compilation_flags):
     flags = list(compilation_flags.host_cxxflags)
-    flags += split(base_cxxflags[0])
+    if base_cxxflags:
+        flags += split(base_cxxflags[0])
     return flags
 
 
 @depends("HOST_LDFLAGS", linker_flags, host_linker_ldflags, host, host_c_compiler)
 @imports(_from="mozbuild.shellutil", _import="split")
 def host_ldflags(env_ldflags, linker_flags, host_linker_ldflags, host, compiler):
-    flags = split(env_ldflags[0])
+    flags = []
+    if env_ldflags:
+        flags += split(env_ldflags[0])
     flags += host_linker_ldflags
     if host.kernel == "WINNT" and compiler.type == "clang-cl":
         if host.cpu == "x86":
             flags += ["-MACHINE:X86"]
         elif host.cpu == "x86_64":
             flags += ["-MACHINE:X64"]
     flags += linker_flags.host_ldflags
     return flags
 
 
 @depends("CPPFLAGS")
 @imports(_from="mozbuild.shellutil", _import="split")
 def os_cppflags(env_cppflags):
-    flags = split(env_cppflags[0])
+    flags = []
+    if env_cppflags:
+        flags = split(env_cppflags[0])
     return flags
 
 
 @depends("CFLAGS", compilation_flags, android_flags, all_arm_flags)
 @imports(_from="mozbuild.shellutil", _import="split")
 def os_cflags(env_cflags, compilation_flags, android_flags, all_arm_flags):
     flags = []
     if android_flags:
         flags.extend(android_flags.cflags)
     if all_arm_flags:
         flags.extend(all_arm_flags)
     flags.extend(compilation_flags.cflags)
-    flags.extend(split(env_cflags[0]))
+    if env_cflags:
+        flags.extend(split(env_cflags[0]))
     return flags
 
 
 @depends("CXXFLAGS", compilation_flags, android_flags, all_arm_flags)
 @imports(_from="mozbuild.shellutil", _import="split")
 def os_cxxflags(env_cxxflags, compilation_flags, android_flags, all_arm_flags):
     flags = []
     if android_flags:
         flags.extend(android_flags.cxxflags)
     if all_arm_flags:
         flags.extend(all_arm_flags)
     flags.extend(compilation_flags.cxxflags)
-    flags.extend(split(env_cxxflags[0]))
+    if env_cxxflags:
+        flags.extend(split(env_cxxflags[0]))
     return flags
 
 
 @depends(
     "ASFLAGS",
     asm_flags,
     android_flags,
     all_arm_flags,
@@ -659,17 +668,18 @@ def os_asflags(
     if android_flags:
         flags.extend(android_flags.asflags)
         flags.append("-DANDROID")
     if all_arm_flags:
         flags.extend(all_arm_flags)
     flags.extend(asm_flags.asflags)
     if build_project != "js" and c_compiler.type != "clang-cl":
         flags.extend(defines_cpp_flags)
-    flags.extend(split(env_asflags[0]))
+    if env_asflags:
+        flags.extend(split(env_asflags[0]))
     return flags
 
 
 # VS2012+ defaults to -arch:SSE2. We want to target nothing more recent, so set
 # that explicitly here unless another target arch has already been set.
 @template
 def add_sse2_flag_if_needed(flags):
     @depends(
diff --git a/build/moz.configure/toolchain.configure b/build/moz.configure/toolchain.configure
--- a/build/moz.configure/toolchain.configure
+++ b/build/moz.configure/toolchain.configure
@@ -116,17 +116,17 @@ def moz_optimize_flags(option, forced_pg
         return [forced_pgo_optimization_level]
 
 
 set_config("MOZ_OPTIMIZE", moz_optimize)
 add_old_configure_assignment("MOZ_OPTIMIZE", moz_optimize)
 add_old_configure_assignment("MOZ_CONFIGURE_OPTIMIZE_FLAGS", moz_optimize_flags)
 
 
-@depends("MOZ_OPTIMIZE_FLAGS")
+@depends_if("MOZ_OPTIMIZE_FLAGS")
 @imports(_from="mozbuild.shellutil", _import="split")
 def moz_optimize_flags(env_optimize_flags):
     return split(env_optimize_flags[0])
 
 
 add_old_configure_assignment("MOZ_OPTIMIZE_FLAGS", moz_optimize_flags)
 
 


